db/view/view_update_generator.cc

```diff
@@ -408,15 +414,36 @@ future<> view_update_generator::generate_and_propagate_view_updates(const replic
         }
         tracing::trace(tr_state, "Generated {} view update mutations", updates->size());
         auto units = seastar::consume_units(_db.view_update_sem(), memory_usage_of(*updates));
-        if (_db.view_update_sem().current() == 0) {
-            // We don't have resources to propagate view updates for this write. If we reached this point, we failed to
-            // throttle the client. The memory queue is already full, waiting on the semaphore would block view updates
-            // that we've already started applying, and generating hints would ultimately result in the disk queue being
-            // full. Instead, we drop the base write, which will create inconsistencies between base replicas, but we
-            // will fix them using repair.
-            err = std::make_exception_ptr(exceptions::overloaded_exception("Too many view updates started concurrently"));
-            break;
+        if (batch_num == 0) {
+            if (_db.view_update_sem().current() == 0) {
+                // We don't have resources to propagate view updates for this write. If we reached this point, we failed to
+                // throttle the client. The memory queue is already full, waiting on the semaphore would block view updates
+                // that we've already started applying, and generating hints would ultimately result in the disk queue being
+                // full. Instead, we drop the base write, which will create inconsistencies between base replicas, but we
+                // will fix them using repair.
+                err = std::make_exception_ptr(exceptions::overloaded_exception("Too many view updates started concurrently"));
+                break;
+            }
         }
+        // To prevent overload we sleep for a moment before sending another batch of view updates.
+        // The amount of time to sleep for is chosen based on how full the view update backlog is,
+        // the more full the queue of pending view updates is the more aggresively we should delay
+        // new ones.
+        // The first batch of updates doesn't have any delays becase it's slowed down by the other throttling mechanism,
+        // the one which limits the number of incoming client requests by delaying the response to the client.
+        if (batch_num > 0) {
+            update_backlog local_backlog = _db.get_view_update_backlog();
+            std::chrono::microseconds throttle_delay =  calculate_view_update_throttling_delay(local_backlog, timeout);
+
+            co_await seastar::sleep(throttle_delay);
+
+            if (db::timeout_clock::now() > timeout) {
+                err = std::make_exception_ptr(exceptions::view_update_generation_timeout_exception());
+                throw exceptions::view_update_generation_timeout_exception();
```

Comment:
How about

```c++
co_await coroutine::return_exception(exceptions::view_update_generation_timeout_exception());
```

This will return the exception from the coroutine without using the costly stack unwinding machinery. It's especially important during overload.